### ISSUE TITLE
Add setting on how to reflect on private members in DefaultRecordMapper

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10341]
Using a `SecurityManager` that prevents reflection on private members was not very popular pre Java 9, but now that we have modularity, running into `InaccessibleObjectException` may happen more often, see e.g. this bug: https://github.com/jOOQ/jOOQ/issues/10267

jOOQ's `DefaultRecordMapper` tries to open up all target types to reflection using `AccessibleObject.setAccessible()`. This will start to cause issues when running jOOQ on the module path. E.g. if the POJO is like this for whatever reason:

```java
public class X {
  public int a;
  private int b;
}
```

We should offer a new setting governing what to do with the private members:

- [ ] Ignore them (except to read annotations)
- [ ] Try to set them and ignore `InaccessibleObjectException` and `SecurityException`
- [ ] Try to set them and fail (current behaviour, and default)

This needs a bit of thought.

**Comment by hubick** [comment]
TL;DR: Option 2.

I could see declaring private members on a data model because you don't want software consuming it to have direct access to those, but may still want jOOQ to populate them from your DB, so ignoring them altogether wouldn't allow for that. One can interpret `private` to mean "inaccessible to anyone not given explicit access via a JPMS open declaration".

The "setting" here could be "is the target module open for jOOQ to reflect on it?" (which has the added benefit of letting one set it for each and every target type independently)

That is, default behaviour of "if jOOQ has been given permission, then go for it, otherwise don't, and that's not an error".  Basically, assume by default that the situation you're in is in fact as the user intended.

The only danger with that proposal is users who _did_ intend to inject private members, but then _failed_ to properly open their target module, and are then confused as to why it's not actually being populated. For that, I would consider adding a debug level warning the first time jOOQ encounters a target class where a record contains a field name which _would_ have been mapped to a private member, but then _wasn't_ due to JPMS access permissions.

Now, that said, I'm not averse to the idea of a "hey jOOQ, stop twiddling with my privates" setting - where public means public and private means private, for everyone, and stop trying to act like you're special and it doesn't apply to you. But, as someone who does like that behaviour, I use @ConstructorProperties, which may convey that intent already.

**Comment by lukaseder** [comment]
> I could see declaring private members on a data model because you don't want software consuming it to have direct access to those, but may still want jOOQ to populate them from your DB, so ignoring them altogether wouldn't allow for that. One can interpret `private` to mean "inaccessible to anyone not given explicit access via a JPMS open declaration".

You don't need JPMS to argue along these lines. JPMS just adds a bit more protection, and more semantics (a virtual `module` or `internal` (as in Kotlin) visibility, if you wish).

I don't see this in any way different than using a `SecurityManager`.

> The "setting" here could be "is the target module open for jOOQ to reflect on it?" (which has the added benefit of letting one set it for each and every target type independently)

I don't want to make this modularity specific, because otherwise, we'd have to make setting that check if things are `protected`, `default` visibility, `private` as well for someone else who would like that granularity, instead of JPMS's (I think that still very few systems actually use JPMS, still), or, again, allow for distinguishing between JPMS or `SecurityManager` caused access denial.

> That is, default behaviour of "if jOOQ has been given permission, then go for it, otherwise don't, and that's not an error". Basically, assume by default that the situation you're in is in fact as the user intended.

Changing defaults may be good for new users, or the few that explicitly agree with the change. For many others, it's a hassle, and there's always something that's being overlooked.

Anyway, there's no rush to implement this. I prefer to collect more user input in this area, first. I think that ultimately, most value is provided by offering users a more composable mapping library: #10348
